File: tests/WinMix.Audio.Tests/ProcessInfoCacheTests.cs
using System.Diagnostics;
using WinMix.Audio;
using Xunit;

namespace WinMix.Audio.Tests;

public class ProcessInfoCacheTests
{
    [Fact]
    public void PidZeroIsLabelledAsSystemSounds()
    {
        var cache = new ProcessInfoCache();

        var info = cache.Get(0);

        Assert.Equal("System sounds", info.FriendlyName);
        Assert.Null(info.ExecutablePath);
    }

    [Fact]
    public void ResolvesTheCurrentProcess()
    {
        var cache = new ProcessInfoCache();
        var pid = (uint)Environment.ProcessId;

        var info = cache.Get(pid);

        Assert.Equal(pid, info.Pid);
        Assert.NotNull(info.ExecutablePath);
        Assert.True(File.Exists(info.ExecutablePath));
        Assert.False(string.IsNullOrWhiteSpace(info.FriendlyName));
    }

    [Fact]
    public void RepeatedLookupsAreServedFromCache()
    {
        var cache = new ProcessInfoCache();
        var pid = (uint)Environment.ProcessId;

        Assert.Same(cache.Get(pid), cache.Get(pid));
    }

    [Fact]
    public void AnUnknownPidStillYieldsADisplayableName()
    {
        var cache = new ProcessInfoCache();

        // Odd values are never valid pids on Windows, so this cannot collide with
        // a live process and race the assertion.
        var info = cache.Get(uint.MaxValue - 1);

        Assert.False(string.IsNullOrWhiteSpace(info.FriendlyName));
        Assert.Null(info.ExecutablePath);
    }

    [Fact]
    public void TrimDropsProcessesThatStoppedPlaying()
    {
        var cache = new ProcessInfoCache();
        var pid = (uint)Environment.ProcessId;
        var first = cache.Get(pid);

        cache.Trim(new HashSet<uint>());

        // A fresh instance proves the entry was evicted rather than reused, which
        // is what keeps a recycled pid from inheriting the old label.
        Assert.NotSame(first, cache.Get(pid));
    }

    [Fact]
    public void TrimKeepsProcessesStillPlaying()
    {
        var cache = new ProcessInfoCache();
        var pid = (uint)Environment.ProcessId;
        var first = cache.Get(pid);

        cache.Trim(new HashSet<uint> { pid });

        Assert.Same(first, cache.Get(pid));
    }

    [Fact]
    public void FriendlyNameIsNeverAFullPath()
    {
        var cache = new ProcessInfoCache();

        var info = cache.Get((uint)Environment.ProcessId);

        Assert.DoesNotContain(Path.DirectorySeparatorChar, info.FriendlyName);
    }
}

File: native/tests/ProcessInfoCacheTests.cpp
#include "doctest.h"
#include "winmix/audio/ProcessInfoCache.h"

#include <windows.h>

#include <cstdint>
#include <filesystem>

using winmix::audio::ProcessInfoCache;

TEST_CASE("PidZeroIsLabelledAsSystemSounds")
{
    ProcessInfoCache cache;

    const auto& info = cache.Get(0);

    CHECK(info.friendlyName == L"System sounds");
    CHECK(!info.executablePath.has_value());
}

TEST_CASE("ResolvesTheCurrentProcess")
{
    ProcessInfoCache cache;
    const uint32_t pid = GetCurrentProcessId();

    const auto& info = cache.Get(pid);

    CHECK(info.pid == pid);
    REQUIRE(info.executablePath.has_value());
    CHECK(std::filesystem::exists(*info.executablePath));
    CHECK(!info.friendlyName.empty());
}

TEST_CASE("RepeatedLookupsAreServedFromCache")
{
    ProcessInfoCache cache;
    const uint32_t pid = GetCurrentProcessId();

    const uint64_t first = cache.Get(pid).resolveSequence;
    const uint64_t second = cache.Get(pid).resolveSequence;

    CHECK(first == second);
}

TEST_CASE("AnUnknownPidStillYieldsADisplayableName")
{
    ProcessInfoCache cache;

    // Not a multiple of 4, so this can never collide with a live Windows pid.
    const auto& info = cache.Get(UINT32_MAX - 1);

    CHECK(!info.friendlyName.empty());
    CHECK(!info.executablePath.has_value());
}

TEST_CASE("TrimDropsProcessesThatStoppedPlaying")
{
    ProcessInfoCache cache;
    const uint32_t pid = GetCurrentProcessId();
    const uint64_t first = cache.Get(pid).resolveSequence;

    cache.Trim({});

    // A different sequence number proves the entry was evicted and
    // re-resolved rather than reused, which is what keeps a recycled pid
    // from inheriting the old label.
    CHECK(cache.Get(pid).resolveSequence != first);
}

TEST_CASE("TrimKeepsProcessesStillPlaying")
{
    ProcessInfoCache cache;
    const uint32_t pid = GetCurrentProcessId();
    const uint64_t first = cache.Get(pid).resolveSequence;

    cache.Trim({pid});

    CHECK(cache.Get(pid).resolveSequence == first);
}

TEST_CASE("FriendlyNameIsNeverAFullPath")
{
    ProcessInfoCache cache;

    const auto& info = cache.Get(GetCurrentProcessId());

    CHECK(info.friendlyName.find(L'\\') == std::wstring::npos);
}
